crates/sui/src/upgrade_compatibility.rs

```diff
@@ -101,7 +109,127 @@ pub(crate) enum UpgradeCompatibilityModeError {
+                    self.print_missing_definition(
+                        "Enum",
+                        name.to_string(),
+                        compiled_unit_with_source,
+                    )?;
+                }
+                UpgradeCompatibilityModeError::FunctionMissingPublic { name, .. } => {
+                    self.print_missing_definition(
+                        "Function",
+                        name.to_string(),
+                        compiled_unit_with_source,
+                    )?;
+                }
+                UpgradeCompatibilityModeError::FunctionMissingEntry { name, .. } => {
+                    self.print_missing_definition(
+                        "Function",
+                        name.to_string(),
+                        compiled_unit_with_source,
+                    )?;
+                }
+                _ => {}
+            }
+        }
+        Ok(())
+    }
+
+    /// retrieve the source, caches the source after the first read
+    fn source(
+        &mut self,
+        compiled_unit_with_source: &CompiledUnitWithSource,
+    ) -> Result<&String, Error> {
+        if self.source.is_none() {
+            let source_path = compiled_unit_with_source.source_path.clone();
+            let source_content = fs::read_to_string(&source_path)?;
+            self.source = Some(source_content);
+        }
+        Ok(self.source.as_ref().unwrap())
+    }
+
+    /// Print missing definition errors, e.g. struct, enum, function
+    fn print_missing_definition(
+        &mut self,
+        declaration_kind: &str,
+        identifier_name: String,
+        compiled_unit_with_source: &CompiledUnitWithSource,
+    ) -> Result<(), Error> {
+        let module_name = compiled_unit_with_source.unit.name;
+        let source_path = compiled_unit_with_source.source_path.to_string_lossy();
+        let source = self.source(compiled_unit_with_source)?;
+
+        let start = compiled_unit_with_source
+            .unit
+            .source_map
+            .definition_location
+            .start() as usize;
+
+        let end = compiled_unit_with_source
+            .unit
+            .source_map
+            .definition_location
+            .end() as usize;
+
+        let mut files = SimpleFiles::new();
+        let file_id = files.add(source_path, &source);
+
+        let diag = Diagnostic::error()
+            .with_message(format!("{} is missing", declaration_kind))
+            .with_labels(vec![Label::primary(file_id, start..end).with_message(
+                format!(
+                    "Module '{}' expected {} '{}', but found none",
+                    declaration_kind, module_name, identifier_name
+                ),
+            )])
+            .with_notes(vec![format!(
+                "The {} is missing in the new module, add the previously defined: '{}'",
+                declaration_kind, identifier_name
+            )]);
```

Comment:
I think the order of parameters on L215 might be off -- I find it's generally easier to make sense of string interpolations if you use the form where the identifier is inline inside the curly braces:

```suggestion
            .with_message(format!("{declaration_kind} is missing"))
            .with_labels(vec![Label::primary(file_id, start..end).with_message(
                format!(
                    "Module '{module_name}' expected {declaration_kind} '{identifer_name}', but found none",
                ),
            )])
            .with_notes(vec![format!(
                "The {declaration_kind} is missing in the new module, add the previously defined: '{identifier_name}'",
            )]);
```